src/core/sock/sockinfo_tcp.cpp

```diff
@@ -894,80 +889,163 @@ ssize_t sockinfo_tcp::tx(xlio_tx_call_attr_t &tx_arg)
-    return false;
+#define TCP_WND_UNAVALABLE(pcb, total_iov_len) false
 #endif
+
+static inline bool is_invalid_iovec(const iovec *iov, size_t sz_iov)
+{
+    return iov == nullptr || sz_iov == 0;
 }
 
 ssize_t sockinfo_tcp::tcp_tx(xlio_tx_call_attr_t &tx_arg)
 {
     iovec *p_iov = tx_arg.attr.iov;
     size_t sz_iov = tx_arg.attr.sz_iov;
-    struct sockaddr *__dst = tx_arg.attr.addr;
-    socklen_t __dstlen = tx_arg.attr.len;
-    int __flags = tx_arg.attr.flags;
+    int flags = tx_arg.attr.flags;
     int errno_tmp = errno;
     int ret = 0;
     int poll_count = 0;
-    uint16_t apiflags = 0;
     err_t err;
-    bool is_send_zerocopy = false;
     void *tx_ptr = NULL;
     struct xlio_pd_key *pd_key_array = NULL;
 
     /* Let allow OS to process all invalid scenarios to avoid any
      * inconsistencies in setting errno values
      */
-    if (unlikely(m_sock_offload != TCP_SOCK_LWIP) || unlikely(!p_iov) || unlikely(0 == sz_iov)) {
-        ret = socket_fd_api::tx_os(tx_arg.opcode, p_iov, sz_iov, __flags, __dst, __dstlen);
+    if (unlikely(m_sock_offload != TCP_SOCK_LWIP) || unlikely(is_invalid_iovec(p_iov, sz_iov))) {
+        struct sockaddr *dst = tx_arg.attr.addr;
+        socklen_t dstlen = tx_arg.attr.len;
+        ret = socket_fd_api::tx_os(tx_arg.opcode, p_iov, sz_iov, flags, dst, dstlen);
         save_stats_tx_os(ret);
         return ret;
     }
 
-retry_is_ready:
+    if (unlikely(!is_connected_and_ready_to_send())) {
+        return -1;
+    }
 
-    if (unlikely(!is_rts())) {
+    si_tcp_logfunc("tx: iov=%p niovs=%d", p_iov, sz_iov);
 
-        if (m_conn_state == TCP_CONN_TIMEOUT) {
-            si_tcp_logdbg("TX timed out");
-            errno = ETIMEDOUT;
-        } else if (m_conn_state == TCP_CONN_CONNECTING) {
-            si_tcp_logdbg("TX while async-connect on socket go to poll");
-            rx_wait_helper(poll_count, false);
-            if (m_conn_state == TCP_CONN_CONNECTED) {
-                goto retry_is_ready;
-            }
-            si_tcp_logdbg("TX while async-connect on socket return EAGAIN");
-            errno = EAGAIN;
-        } else if (m_conn_state == TCP_CONN_RESETED) {
-            si_tcp_logdbg("TX on reseted socket");
-            errno = ECONNRESET;
-        } else if (m_conn_state == TCP_CONN_ERROR) {
-            si_tcp_logdbg("TX on connection failed socket");
-            errno = ECONNREFUSED;
-        } else {
-            si_tcp_logdbg("TX on disconnected socket");
-            errno = EPIPE;
+    if (m_sysvar_rx_poll_on_tx_tcp) {
+        rx_wait_helper(poll_count, false);
+    }
+
+    bool is_dummy = IS_DUMMY_PACKET(flags);
+    bool is_blocking = BLOCK_THIS_RUN(m_b_blocking, flags);
+    bool is_packet_zerocopy = (flags & MSG_ZEROCOPY) && ((m_b_zc) || (tx_arg.opcode == TX_FILE));
+    if (unlikely(is_dummy) || unlikely(!is_packet_zerocopy) || unlikely(is_blocking)) {
+        return tcp_tx_slow_path(tx_arg);
+    }
+
+    bool is_send_zerocopy = tx_arg.opcode != TX_FILE;
```

Comment:
we need to rename this var to is_user_zc or something like that. becuase file is also zero copy just not from user.

Reply:
done